File: modules/foundation/src/main/java/com/projecthub/base/milestone/domain/entity/Milestone.java
package com.projecthub.base.milestone.domain.entity;

import com.projecthub.base.milestone.domain.enums.MilestoneStatus;
import com.projecthub.base.project.domain.entity.Project;
import com.projecthub.base.shared.domain.entity.BaseEntity;
import com.projecthub.base.task.domain.entity.Task;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

/**
 * Represents a project milestone with associated tasks and dependencies.
 *
 * <p>
 * This entity includes details such as name, description, due date, status,
 * progress, and relationships to projects, tasks, and other milestones.
 * </p>
 *
 * <p>
 * Key functionalities:
 * <ul>
 * <li>Check if the milestone is overdue.</li>
 * <li>Update progress and handle status transitions.</li>
 * <li>Manage dependencies between milestones.</li>
 * <li>Start, complete, block, or cancel the milestone.</li>
 * <li>Determine if the milestone is active or blocked.</li>
 * </ul>
 * </p>
 *
 * <p>
 * Constraints:
 * <ul>
 * <li>Name must be between 3 and 100 characters and not blank.</li>
 * <li>Description can be up to 500 characters.</li>
 * <li>Due date must be in the future.</li>
 * <li>Progress must be between 0 and 100.</li>
 * <li>Cannot have circular dependencies.</li>
 * <li>Cannot complete if there are incomplete tasks.</li>
 * <li>Cannot block or cancel if already completed or cancelled.</li>
 * </ul>
 * </p>
 *
 * @author
 */
@Entity
@Table(name = "milestones")
@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(callSuper = true, onlyExplicitlyIncluded = true)
@ToString(callSuper = true, onlyExplicitlyIncluded = true)
public class Milestone extends BaseEntity {

    @ToString.Include
    @NotBlank(message = "Name is required")
    @Size(min = 3, max = 100)
    private String name;

    @Size(max = 500)
    private String description;

    @NotNull(message = "Due date is required")
    @Future(message = "Due date must be in the future")
    private LocalDate dueDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @Enumerated(EnumType.STRING)
    @Setter(AccessLevel.PROTECTED)
    private MilestoneStatus status = MilestoneStatus.PENDING;

    @Version
    private Long version;

    @Min(0)
    @Max(100)
    private Integer progress = 0;

    @OneToMany(mappedBy = "milestone", cascade = CascadeType.ALL)
    private Set<Task> tasks = new HashSet<>();

    @ManyToMany
    @JoinTable(name = "milestone_dependencies", joinColumns = @JoinColumn(name = "milestone_id"), inverseJoinColumns = @JoinColumn(name = "dependency_id"))
    private Set<Milestone> dependencies = new HashSet<>();
    private MilestoneStatus targetStatus;

    public boolean isOverdue() {
        return status != MilestoneStatus.COMPLETED && dueDate.isBefore(LocalDate.now());
    }

    public void updateProgress(int newProgress) {
        if (newProgress < 0 || newProgress > 100) {
            throw new IllegalArgumentException("Progress must be between 0 and 100");
        }
        this.progress = newProgress;
        if (newProgress == 100) {
            this.status = MilestoneStatus.COMPLETED;
        }
    }

    public void addDependency(Milestone dependency) {
        if (dependency.equals(this)) {
            throw new IllegalArgumentException("Milestone cannot depend on itself");
        }
        dependencies.add(dependency);
    }

    public boolean canStart() {
        return dependencies.stream()
            .allMatch(m -> m.status == MilestoneStatus.COMPLETED || m.status == MilestoneStatus.CANCELLED);
    }

    public void start() {
        if (!canStart()) {
            throw new IllegalStateException("Dependencies not met");
        }
        if (status != MilestoneStatus.PENDING) {
            throw new IllegalStateException("Can only start pending milestones");
        }
        status = MilestoneStatus.IN_PROGRESS;
    }

    public boolean isBlocked() {
        return status == MilestoneStatus.BLOCKED;
    }

    public void complete() {
        if (!tasks.isEmpty() && tasks.stream().anyMatch(t -> !t.isCompleted())) {
            throw new IllegalStateException("Cannot complete milestone with incomplete tasks");
        }
        status = MilestoneStatus.COMPLETED;
        progress = 100;
    }

    public void block() {
        if (status == MilestoneStatus.COMPLETED || status == MilestoneStatus.CANCELLED) {
            throw new IllegalStateException("Cannot block completed or cancelled milestone");
        }
        status = MilestoneStatus.BLOCKED;
    }

    public void cancel() {
        if (status == MilestoneStatus.COMPLETED) {
            throw new IllegalStateException("Cannot cancel completed milestone");
        }
        status = MilestoneStatus.CANCELLED;
    }

    public boolean isActive() {
        return status == MilestoneStatus.IN_PROGRESS;
    }

    public boolean hasIncompleteTasks() {
        return tasks.stream().anyMatch(task -> !task.isCompleted());
    }

    public boolean hasCyclicDependencies() {
        Set<Milestone> visited = new HashSet<>();
        Set<Milestone> recursionStack = new HashSet<>();
        return checkCyclicDependencies(this, visited, recursionStack);
    }

    private boolean checkCyclicDependencies(Milestone milestone,
                                            Set<Milestone> visited,
                                            Set<Milestone> recursionStack) {
        if (recursionStack.contains(milestone)) {
            return true;
        }
        if (visited.contains(milestone)) {
            return false;
        }

        visited.add(milestone);
        recursionStack.add(milestone);

        boolean hasCycle = milestone.dependencies.stream()
            .anyMatch(dep -> checkCyclicDependencies(dep, visited, recursionStack));

        recursionStack.remove(milestone);
        return hasCycle;
    }

    public boolean isCompleted() {
        return status == MilestoneStatus.COMPLETED;
    }

    public MilestoneStatus getTargetStatus() {
        return targetStatus;
    }

    public void setTargetStatus(MilestoneStatus status) {
        this.targetStatus = status;
    }
}

File: src/main/kotlin/com/projecthub/domain/milestone/Milestone.kt
package com.projecthub.domain.milestone

import com.projecthub.domain.AggregateRoot
import com.projecthub.domain.milestone.event.MilestoneCreatedEvent
import com.projecthub.domain.milestone.event.MilestoneCompletedEvent
import com.projecthub.domain.milestone.event.MilestoneDueDateChangedEvent
import com.projecthub.domain.milestone.event.MilestoneAssignedEvent
import com.projecthub.domain.task.Task
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*

class Milestone private constructor(
    val id: String,
    var name: String,
    var description: String,
    var dueDate: LocalDate,
    var status: MilestoneStatus = MilestoneStatus.PENDING,
    var progress: Int = 0,
    var projectId: String,
    var assigneeId: String? = null,
    var completionDate: LocalDate? = null,
    var createdAt: LocalDateTime = LocalDateTime.now(),
    var updatedAt: LocalDateTime = LocalDateTime.now()
) : AggregateRoot<String>() {

    override fun getId(): String = id

    private val _tasks: MutableSet<Task> = mutableSetOf()
    val tasks: Set<Task> get() = _tasks.toSet()

    private val _dependencies: MutableSet<Milestone> = mutableSetOf()
    val dependencies: Set<Milestone> get() = _dependencies.toSet()

    val isCompleted: Boolean get() = status == MilestoneStatus.COMPLETED
    val isBlocked: Boolean get() = status == MilestoneStatus.BLOCKED

    companion object {
        /**
         * Creates a new Milestone instance and registers a milestone created event
         */
        fun create(
            id: String,
            name: String,
            description: String,
            projectId: String,
            dueDate: LocalDate
        ): Milestone {
            val milestone = Milestone(
                id = id,
                name = name,
                description = description,
                dueDate = dueDate,
                projectId = projectId
            )

            // Register domain event for new milestone creation
            milestone.registerEvent(
                MilestoneCreatedEvent(
                    milestoneId = id,
                    name = name,
                    projectId = projectId,
                    dueDate = dueDate.toString(),
                    description = description
                )
            )

            return milestone
        }
    }

    fun updateStatus(newStatus: MilestoneStatus) {
        validateStatusTransition(newStatus)
        status = newStatus
        updatedAt = LocalDateTime.now()

        if (newStatus == MilestoneStatus.COMPLETED) {
            completionDate = LocalDate.now()
            progress = 100

            registerEvent(
                MilestoneCompletedEvent(
                    milestoneId = id,
                    projectId = projectId,
                    completionDate = completionDate.toString()
                )
            )
        }
    }

    fun addTask(task: Task) {
        _tasks.add(task)
        updateProgress()
        updatedAt = LocalDateTime.now()
    }

    fun assignTo(assigneeId: String) {
        this.assigneeId = assigneeId
        updatedAt = LocalDateTime.now()

        registerEvent(
            MilestoneAssignedEvent(
                milestoneId = id,
                projectId = projectId,
                assigneeId = assigneeId
            )
        )
    }

    fun changeDueDate(newDueDate: LocalDate) {
        val oldDueDate = this.dueDate
        this.dueDate = newDueDate
        updatedAt = LocalDateTime.now()

        registerEvent(
            MilestoneDueDateChangedEvent(
                milestoneId = id,
                projectId = projectId,
                oldDueDate = oldDueDate.toString(),
                newDueDate = newDueDate.toString()
            )
        )
    }

    fun addDependency(dependency: Milestone) {
        require(!hasCyclicDependencies(dependency)) { "Adding this dependency would create a cycle" }
        require(!dependency.dueDate.isAfter(this.dueDate)) { "Dependency due date must be before milestone due date" }
        _dependencies.add(dependency)
        updatedAt = LocalDateTime.now()
    }

    fun updateProgress() {
        val completedTasks = tasks.count { it.isCompleted }
        val newProgress = if (tasks.isEmpty()) 0
        else ((completedTasks.toDouble() / tasks.size) * 100).toInt()

        require(!(status == MilestoneStatus.COMPLETED && newProgress != 100)) {
            "Completed milestone must have 100% progress"
        }

        progress = newProgress
        updatedAt = LocalDateTime.now()
    }

    private fun validateStatusTransition(newStatus: MilestoneStatus) {
        val validTransitions = when (status) {
            MilestoneStatus.PENDING -> setOf(
                MilestoneStatus.IN_PROGRESS,
                MilestoneStatus.BLOCKED,
                MilestoneStatus.CANCELLED
            )

            MilestoneStatus.IN_PROGRESS -> setOf(
                MilestoneStatus.COMPLETED,
                MilestoneStatus.BLOCKED,
                MilestoneStatus.CANCELLED
            )

            MilestoneStatus.BLOCKED -> setOf(
                MilestoneStatus.PENDING,
                MilestoneStatus.CANCELLED
            )

            MilestoneStatus.COMPLETED -> emptySet()
            MilestoneStatus.CANCELLED -> emptySet()
        }

        require(validTransitions.contains(newStatus)) {
            "Invalid status transition from $status to $newStatus"
        }

        when (newStatus) {
            MilestoneStatus.IN_PROGRESS -> {
                require(dependencies.all { it.isCompleted }) {
                    "Cannot start milestone with incomplete dependencies"
                }
            }

            MilestoneStatus.COMPLETED -> {
                require(tasks.all { it.isCompleted }) {
                    "Cannot complete milestone with incomplete tasks"
                }
                require(progress == 100) {
                    "Cannot complete milestone with progress less than 100%"
                }
            }

            MilestoneStatus.BLOCKED -> {
                val hasBlockedDependency = dependencies.any { it.isBlocked }
                val hasIncompleteDependency = dependencies.any { !it.isCompleted }
                val hasUnassignedTasks = tasks.any { it.assignee == null }
                require(hasBlockedDependency || hasIncompleteDependency || hasUnassignedTasks) {
                    "Cannot block milestone without blocked/incomplete dependencies or unassigned tasks"
                }
            }

            MilestoneStatus.CANCELLED -> {
                require(!isCompleted) { "Cannot cancel completed milestone" }
            }

            MilestoneStatus.PENDING -> {
                require(status == MilestoneStatus.BLOCKED) {
                    "Can only transition to PENDING from BLOCKED status"
                }
                require(dependencies.all { it.isCompleted }) {
                    "All dependencies must be completed before transitioning to PENDING"
                }
            }
        }
    }

    private fun hasCyclicDependencies(newDependency: Milestone, visited: MutableSet<String> = mutableSetOf()): Boolean {
        if (!visited.add(this.id)) return true
        if (this == newDependency) return true

        return dependencies.any { it.hasCyclicDependencies(newDependency, visited) }
    }
}

enum class MilestoneStatus {
    PENDING,
    IN_PROGRESS,
    BLOCKED,
    COMPLETED,
    CANCELLED
}
